model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/connections/DefaultHotRodConnectionProviderFactory.java

```diff
@@ -111,32 +119,52 @@ public void lazyInit() {
-        } else if (reindexCaches != null && !reindexCaches.isEmpty()){
-            Arrays.stream(reindexCaches.split(","))
-                .map(String::trim)
-                    .filter(e -> !e.isEmpty())
-                    .filter(remoteCaches::contains)
-                    .peek(cacheName -> LOG.infof("Reindexing %s cache. This can take a long time to complete. While the rebuild operation is in progress, queries might return fewer results.", cacheName))
-                    .peek(remoteCacheManager::getCache) // access the caches to force their creation, otherwise reindexing fails if cache doesn't exist
-                    .forEach(administration::reindexCache);
+        // Acquire initial phase lock to avoid concurrent schema update
+        RemoteCache<String, String> locksCache = remoteCacheManager.getCache(HOT_ROD_LOCKS_CACHE_NAME);
+        try {
+            ConcurrentMapLocksUtils.repeatPutIfAbsent(() -> locksCache.withFlags(Flag.FORCE_RETURN_VALUE), HOT_ROD_INIT_LOCK_NAME, 900, 50);
+
+            Set<String> remoteCaches = ENTITY_DESCRIPTOR_MAP.values().stream()
+                    .map(HotRodEntityDescriptor::getCacheName).collect(Collectors.toSet());
+
+            LOG.debugf("Uploading proto schema to Infinispan server.");
+            registerSchemata();
+
+            String reindexCaches = config.get("reindexCaches", null);
+            RemoteCacheManagerAdmin administration = remoteCacheManager.administration();
+            if (reindexCaches != null && reindexCaches.equals("all")) {
+                LOG.infof("Reindexing all caches. This can take a long time to complete. While the rebuild operation is in progress, queries might return fewer results.");
+                remoteCaches.stream()
+                        .peek(remoteCacheManager::getCache) // access the caches to force their creation, otherwise reindexing fails if cache doesn't exist
+                        .forEach(administration::reindexCache);
+            } else if (reindexCaches != null && !reindexCaches.isEmpty()) {
+                Arrays.stream(reindexCaches.split(","))
+                        .map(String::trim)
+                        .filter(e -> !e.isEmpty())
+                        .filter(remoteCaches::contains)
+                        .peek(cacheName -> LOG.infof("Reindexing %s cache. This can take a long time to complete. While the rebuild operation is in progress, queries might return fewer results.", cacheName))
+                        .peek(remoteCacheManager::getCache) // access the caches to force their creation, otherwise reindexing fails if cache doesn't exist
+                        .forEach(administration::reindexCache);
+            }
+
+            if (!ConcurrentMapLocksUtils.removeWithInstanceIdentifier(locksCache, HOT_ROD_INIT_LOCK_NAME)) {
+                throw new RuntimeException("Cannot release HotRod init lock");
+            }
+            LOG.infof("HotRod client configuration was successful.");
```

Comment:
Should this be in `finally` block?

Reply:
HotRod initialization is not successful if this line is skipped due to an exception. Why do you think this should be in `finally`?